keps/sig-cluster-lifecycle/kubeadm/20190916-kubeadm-operator.md

```diff
@@ -0,0 +1,359 @@
+- [x] KEP approvers have set the KEP status to `implementable`
+- [x] Design details are appropriately documented
+- [ ] Test plan is in place, giving consideration to SIG Architecture and SIG Testing input
+- [ ] Graduation criteria is in place
+- [ ] "Implementation History" section is up-to-date for milestone
+- [ ] User-facing documentation has been created in [kubernetes/website], for publication to [kubernetes.io]
+- [ ] Supporting documentation e.g., additional design documents, links to mailing list discussions/SIG meetings, relevant PRs/issues, release notes
+
+## Summary
+
+Kubeadm operator would like to enable declarative control of kubeadm workflows, automating the execution and the orchestration of such tasks across existing nodes in a cluster.
+
+## Motivation
+
+Kubeadm binary can execute actions only on the machine where it is running e.g. it is not possible to execute actions on other nodes, to copy files across nodes, etc.
+
+As a consequence, most of the kubeadm workflows, like e.g. [kubeadm upgrade](https://kubernetes.io/docs/reference/setup-tools/kubeadm/kubeadm-upgrade/), consists of a complex sequence of tasks that should be manually executed/orchestrated across all the existing nodes in the cluster.
+
+Such a user experience is not ideal, potentially error-prone, and this approach can be considered also a limitation/blocker for implementing more complex workflows like e.g. rotating certificate authorities, change the settings of an existing cluster etc.
+
+This KEP aims to address such problems by applying the operator pattern to kubeadm workflows. 
+
+### Goals
+
+- To allow declarative control of kubeadm action that leads to "in place" mutations[1] of kubeadm generated artifacts. More specifically this initially targets includes kubeadm workflows:
+  - kubeadm upgrade
+  - certificate renewal
+  - change settings/configuration in an existing cluster (NEW) 
+  - certificate authority rotation (NEW)
+- To automate the orchestration of such mutations across the existing nodes in a cluster
+
+> [1] Please note that we are referring to "in place" mutations of kubeadm generated artifacts in order to highlight the difference between the kubeadm operator and other SCL projects like [Cluster API](https://cluster-api.sigs.k8s.io/), which instead assume node/machines immutable
+
+Considering the complexity of this topic, this document is expected to be subject to some iterations. The goal of the current iteration is to:
+
+- Get initial approval on Summary and Motivation paragraphs
+- To identify a semantic for defining “Actions” to be performed by the kubeadm-operator.
+- To define how the kubeadm-operator should manage kubeadm workflows.
+- To define how the users should interact with the kubeadm-operator, including also observability and error handling.
+- To define how the kubeadm-operator should be deployed in a Kubernetes cluster.
+
+### Non-Goals
```

Comment:
I would explicitly mention that replacing `kubeadm` "raw" workflows is a non-goal, and if something, an enabler for the operator. If more actions are added, `kubeadm` will get those actions as well, and the operator will rely on them, but the user will always be able to run these commands in isolation in a manual fashion.

Reply:
Done!